[SPARK-38785][PYTHON][SQL] Implement ExponentialMovingWindow

zhengruifeng opened the pull request:
### What changes were proposed in this pull request?

initial impl Series.ewm and DataFrame.ewm


| | supported | unsupported |
|---|---|---|
|function|mean|sum/std/var/cov/corr|
|params|com/span/halflife/alpha/min_periods|adjust/ignore_na/axis/method|

other unsupoorted functionality: 

- chaining with groupby: `df.groupby('s1').ewm(com=0.5).mean()`

- `DatetimeIndex`

- do not support dataset containing NULL and NaN for now

### Why are the changes needed?
to support more pandas API


### Does this PR introduce _any_ user-facing change?
yes, new method is added:

```
In [3]: psdf.ewm(com=0.1).mean()
                                                                                
         s1        s2
0  0.200000  2.000000
1  0.016667  1.083333
2  0.547368  2.827068
3  0.231557  1.165984
4  0.384688  0.105992
5  0.489517  0.009636
6  0.589956  0.000876

```


### How was this patch tested?
added testsuits

zhengruifeng commented:
```
scala> spark.sql(""" SELECT a, b, c, ewm(c, 0.1) OVER (PARTITION BY b ORDER BY a) FROM VALUES ('A', 0, 2), ('B', 1, 1), ('C', 0, 3), ('D', 1, 1) tab(a, b, c); """).show

+---+---+---+--------------------------------------------------------------------------------------------------------+
|  a|  b|  c|ewm(c) OVER (PARTITION BY b ORDER BY a ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)|
+---+---+---+--------------------------------------------------------------------------------------------------------+
|  A|  0|  2|                                                                                                     2.0|
|  C|  0|  3|                                                                                       2.526315789473684|
|  B|  1|  1|                                                                                                     1.0|
|  D|  1|  1|                                                                                                     1.0|
+---+---+---+--------------------------------------------------------------------------------------------------------+
```

zhengruifeng committed:
basic impl

zhengruifeng committed:
add data check

zhengruifeng commented:
this PR is still WIP, but a basic funcation was implemented.

I personaly thought it is not feasible to efficiently compute `ewm` for each x_t with existing window functions

![image](https://user-images.githubusercontent.com/7322292/161919619-4dadafc6-ed9c-4c7b-8f07-63e1a7eb55d7.png)

so a new window function is added to compute ewm.mean.

However, I found that I should add a `public` api in the SQL side, otherwise, the PySpark side can not find it. Are there some methods to work around it? I think we should not expose it to end users of SparkSQL.

@HyukjinKwon @ueshin could you please take a glance? Am I on the right way? Thanks

zhengruifeng commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -1178,6 +1178,8 @@ object functions {
    */
   def row_number(): Column = withExpr { RowNumber() }
 
+  def ewm(e: Column, alpha: Double): Column = withExpr { EWM(e.expr, alpha) }
```
can we make it only visiable to pyspark in some way?

zhengruifeng committed:
move new func to PythonSQLUtils

zhengruifeng committed:
nit

zhengruifeng committed:
add ut

zhengruifeng committed:
more ut

zhengruifeng committed:
reuse Rolling._apply_as_series_or_frame

zhengruifeng committed:
make min_periods optional int

HyukjinKwon commented:
cc @ueshin, @itholic and @xinrong-databricks if you guys find some time to review.

itholic commented on python/pyspark/pandas/window.py:
```diff
@@ -1749,6 +1752,177 @@ def var(self) -> FrameLike:
+            opt_count += 1
+
+        if alpha is not None:
+            if alpha <= 0 or alpha > 1:
```
nit: Maybe `if 0 < alpha <= 1:` for better readability ? Not strong felling, though

itholic commented on python/pyspark/pandas/window.py:
```diff
@@ -1749,6 +1752,177 @@ def var(self) -> FrameLike:
+            opt_count += 1
+
+        if opt_count == 0:
+            raise ValueError("Must pass one of comass, span, halflife, or alpha")
```
Maybe `comass` -> `com` ?? I think it's a typo in pandas.

zhengruifeng replied:
yep. I copied this error msg from pandas, it should be a typo in pandas.

itholic commented on python/pyspark/pandas/window.py:
```diff
@@ -1749,6 +1752,177 @@ def var(self) -> FrameLike:
+
+        See Also
+        --------
+        Series.expanding : Calling object with Series data.
+        DataFrame.expanding : Calling object with DataFrames.
+        Series.mean : Equivalent method for Series.
+        DataFrame.mean : Equivalent method for DataFrame.
+
+        Examples
+        --------
+        The below examples will show expanding mean calculations with window sizes of
+        two and three, respectively.
```
Maybe those are mistakenly from `expanding.mean` ?

zhengruifeng replied:
yes, good catch.  Will update those doc soon. Thanks!

zhengruifeng commented on python/pyspark/pandas/generic.py:
```diff
@@ -2619,6 +2619,56 @@ def expanding(self: FrameLike, min_periods: int = 1) -> "Expanding[FrameLike]":
+
+        min_periods : int, default 0
+            Minimum number of observations in window required to have a value
+            (otherwise result is NA).
```
`(otherwise result is NA)` this line was copid from `rolling` and `expanding` in `generic.py`.

zhengruifeng replied:
let's wait for others feedbacks.

HyukjinKwon replied:
It's copied from pandas' right? Let's leave it as is for now.

itholic replied:
It's not copied pandas but copied from:

https://github.com/apache/spark/blob/2f122ba6d13ea26411fa4bf3e636ced449a8a205/python/pyspark/pandas/generic.py#L2584-L2586

and:

https://github.com/apache/spark/blob/2f122ba6d13ea26411fa4bf3e636ced449a8a205/python/pyspark/pandas/generic.py#L2610-L2612

In our code base.

pandas says "otherwise, result is `np.nan`." as beolw:

<img width="633" alt="Screen Shot 2022-04-11 at 2 15 34 PM" src="https://user-images.githubusercontent.com/44108233/162669417-3fb30a27-8d29-45af-9af5-06681a24ff86.png">


Maybe we should document as `np.nan` and also fix the document at `rolling` and `expanding` ??

zhengruifeng replied:
what about refining the documnet in another PR? then they can be changed in a batch

HyukjinKwon replied:
Yeah let's do that

zhengruifeng commented on python/pyspark/pandas/tests/test_ewm.py:
```diff
@@ -0,0 +1,126 @@
+            ExponentialMoving(1, 2)
+
+        with self.assertRaisesRegex(ValueError, "min_periods must be >= 0"):
+            ps.range(10).ewm(min_periods=-1, alpha=0.5)
```
nice, will update it.

zhengruifeng commented on python/pyspark/pandas/generic.py:
```diff
@@ -2619,6 +2619,56 @@ def expanding(self: FrameLike, min_periods: int = 1) -> "Expanding[FrameLike]":
+        """
+        Provide exponentially weighted window transformations.
+
+        .. note:: 'min_periods' in pandas-on-Spark works as a fixed window size unlike pandas.
```
this usage of `pandas-on-Spark` was also copied from `rolling` and `expanding` in `generic.py`. I am netural on it.

itholic replied:
Oh, I see.

Then let's just keep it as is for now.

HyukjinKwon replied:
I think it's fine for now.

zhengruifeng committed:
address comments and avoid error for empty constructor

reformat

add missing return type annotation

remove ewm from missing.frame.py

remove ewm from missing.series.py

zhengruifeng commented on python/pyspark/pandas/window.py:
```diff
@@ -1749,6 +1752,188 @@ def var(self) -> FrameLike:
+            Window.unboundedPreceding, Window.currentRow
+        )
+
+        if (com is not None) and (not com >= 0):
```
I found that using `(not com >= 0)` instead of `com < 0` can also check whether `com` is `np.nan`

xinrong-databricks replied:
Good finding!

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:
```diff
@@ -1014,3 +1014,54 @@ case class PercentRank(children: Seq[Expression]) extends RankLike with SizeBase
+ * For now, only function mean is supported. Other functions like sum/var will be added
+ * in the future.
+ */
+private[sql] case class EWM(input: Expression, alpha: Double)
```
We can remove this per SPARK-16813

```suggestion
case class EWM(input: Expression, alpha: Double)
```

HyukjinKwon reviewed (approved):
This is amazing. Took a cursory look and seems fine.

HyukjinKwon commented on python/pyspark/pandas/window.py:
```diff
@@ -1749,6 +1752,188 @@ def var(self) -> FrameLike:
+
+        self._window = window
+        # This unbounded Window is later used to handle 'min_periods' for now.
+        self._unbounded_window = Window.orderBy(NATURAL_ORDER_COLUMN_NAME).rowsBetween(
```
Likewise, this is actually pretty critical that we can't specify partitions while I have other better ideas. We should better document it somewhere that the performance is expected to be slow, and OOM can happen.

zhengruifeng replied:
currently it is documented in the Note part of `ExponentialMoving.mean`.

HyukjinKwon commented on python/pyspark/pandas/window.py:
```diff
@@ -18,16 +18,19 @@
 from functools import partial
 from typing import Any, Callable, Generic, List, Optional
```
```suggestion
from typing import Any, Callable, Generic, List, Optional

import numpy as np
```

HyukjinKwon commented on python/pyspark/pandas/window.py:
```diff
@@ -18,16 +18,19 @@
 )
 
 # For running doctests and reference resolution in PyCharm.
+import numpy as np
```
```suggestion
```

HyukjinKwon commented on python/pyspark/pandas/missing/window.py:
```diff
@@ -124,3 +142,16 @@ class MissingPandasLikeRollingGroupby:
     ndim = _unsupported_property_rolling("ndim")
+
+
+class MissingPandasLikeExponentialMoving:
```
We will have to test it at `python/pyspark/pandas/tests/test_window.py`.

itholic replied:
Maybe you could refer to https://github.com/apache/spark/pull/36034/files when adding the missing test.

HyukjinKwon commented on python/pyspark/pandas/generic.py:
```diff
@@ -2619,6 +2619,56 @@ def expanding(self: FrameLike, min_periods: int = 1) -> "Expanding[FrameLike]":
         return Expanding(self, min_periods=min_periods)
 
+    # TODO: 'adjust', 'ignore_na', 'axis', 'method' parameter should be implemented.
+    def ewm(
```
Would have to list it in https://github.com/apache/spark/blob/master/python/docs/source/reference/pyspark.pandas/window.rst so it's documented.

zhengruifeng committed:
address comments

zhengruifeng commented:
retest this please

zhengruifeng commented:
retest this please

zhengruifeng committed:
fix a typo

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:
```diff
@@ -1014,3 +1015,51 @@ case class PercentRank(children: Seq[Expression]) extends RankLike with SizeBase
+ *    y_t = \frac{\sum_{i=0}^t w_i x_{t-i}}{\sum_{i=0}^t w_i},
+ * where x_t is the input, y_t is the result and the w_i are the weights.
+ */
+@DeveloperApi
+@Experimental
+@Unstable
```
Let's remove these since this class itself isn't exposed as an API.

```suggestion
```

zhengruifeng replied:
ok

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:
```diff
@@ -1014,3 +1015,51 @@ case class PercentRank(children: Seq[Expression]) extends RankLike with SizeBase
+    val beta = Literal(1.0 - alpha)
+    val casted = input.cast(DoubleType)
+    // TODO: after adding param ignore_na, we can remove this check
+    val error = RaiseError(Literal("Input values Must not be Null or NaN")).cast(DoubleType)
```
```suggestion
    val error = RaiseError(Literal("Input values must not be null or NaN")).cast(DoubleType)
```

HyukjinKwon reviewed (approved):
LGTM

zhengruifeng committed:
Update sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

zhengruifeng committed:
Update sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

zhengruifeng committed:
remove unused imports

zhengruifeng commented:
all tests passed  https://github.com/zhengruifeng/spark/actions/runs/2158560305

HyukjinKwon commented:
Merged to master.

zhengruifeng commented:
Thanks all!
